R/module_nested_tabs.R

```diff
@@ -84,235 +84,238 @@ ui_nested_tabs.teal_modules <- function(id, modules, datasets, depth = 0L, is_mo
-#' @rdname module_nested_tabs
-srv_nested_tabs <- function(id, datasets, modules, is_module_specific = FALSE,
-                            reporter = teal.reporter::Reporter$new()) {
+#' @rdname module_teal_module
+srv_teal_module <- function(id,
+                            data_rv,
+                            modules,
+                            datasets = NULL,
+                            slices_global = reactiveVal(teal_slices()),
+                            reporter = teal.reporter::Reporter$new(),
+                            is_active = reactive(TRUE)) {
+  checkmate::assert_string(id)
+  checkmate::assert_class(data_rv, "reactive")
   checkmate::assert_multi_class(modules, c("teal_modules", "teal_module"))
+  checkmate::assert_class(datasets, "reactive", null.ok = TRUE)
+  checkmate::assert_class(slices_global, "reactiveVal")
+  checkmate::assert_class(isolate(slices_global()), "modules_teal_slices")
   checkmate::assert_class(reporter, "Reporter")
-  UseMethod("srv_nested_tabs", modules)
+  UseMethod("srv_teal_module", modules)
 }
 
-#' @rdname module_nested_tabs
+#' @rdname module_teal_module
 #' @export
-srv_nested_tabs.default <- function(id, datasets, modules, is_module_specific = FALSE,
-                                    reporter = teal.reporter::Reporter$new()) {
+srv_teal_module.default <- function(id,
+                                    data_rv,
+                                    modules,
+                                    datasets = NULL,
+                                    slices_global = reactiveVal(teal_slices()),
+                                    reporter = teal.reporter::Reporter$new(),
+                                    is_active = reactive(TRUE)) {
   stop("Modules class not supported: ", paste(class(modules), collapse = " "))
 }
 
-#' @rdname module_nested_tabs
+#' @rdname module_teal_module
 #' @export
-srv_nested_tabs.teal_modules <- function(id, datasets, modules, is_module_specific = FALSE,
-                                         reporter = teal.reporter::Reporter$new()) {
-  checkmate::assert_list(datasets, types = c("list", "FilteredData"))
-
+srv_teal_module.teal_modules <- function(id,
+                                         data_rv,
+                                         modules,
+                                         datasets = NULL,
+                                         slices_global = reactiveVal(teal_slices()),
+                                         reporter = teal.reporter::Reporter$new(),
+                                         is_active = reactive(TRUE)) {
   moduleServer(id = id, module = function(input, output, session) {
-    logger::log_trace("srv_nested_tabs.teal_modules initializing the module { deparse1(modules$label) }.")
+    logger::log_trace("srv_teal_module.teal_modules initializing the module { deparse1(modules$label) }.")
 
-    labels <- vapply(modules$children, `[[`, character(1), "label")
-    modules_reactive <- sapply(
+    modules_output <- sapply(
```

Comment:
Why use `sapply` instead of `lapply`, do we need names? _(from what I tested we don't)_

The `lapply`/`sapply` could be returned immediately

Reply:
We need names, and returned list is named.

> (from what I tested we don't)

What did you tested?

Reply:
Why do we need names? It is being called by `srv_teal` which in turn calls `srv_teal_module`. I'm not seeing the object being referenced via `[[` or other getter, but I may missing something :sweat_smile: 

I've tested with a couple of example applications without any visible change.

<details>

<summary>For example:</summary>

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("teal")

# Tranformer definition (reused in 2 modules) ---------------------------------

transformer_dummy <- teal_data_module(
  label = "Dummy",
  ui = function(id) div("(does nothing)"),
  server = function(id, data) {
    moduleServer(id, function(input, output, session) data)
  }
)

transformer_merge_anl <- teal_data_module(
  label = "Merge ANL",
  ui = function(id) {
    ns <- NS(id)
    tagList(
      div("UI for merge transform"),
      teal.widgets::optionalSelectInput(ns("merge_a"), "Merge A", choices = NULL),
      teal.widgets::optionalSelectInput(ns("merge_b"), "Merge B", choices = NULL)
    )
  },
  server = function(id, data) {
    checkmate::assert_class(data, "reactive")
    moduleServer(id, function(input, output, session) {
      iv <- shinyvalidate::InputValidator$new()
      iv$add_rule("merge_a", shinyvalidate::sv_required("Please select dataset A"))
      iv$add_rule("merge_b", shinyvalidate::sv_required("Please select dataset B"))
      iv$enable()

      reactive_datanames <- reactive({
        req(data())
        teal.data::datanames(data())
      })
      observeEvent(reactive_datanames(), {
        selected_a <- isolate(input$merge_a)
        if (identical(selected_a, "")) selected_a <- restoreInput(session$ns("merge_a"), NULL)
        teal.widgets::updateOptionalSelectInput(
          session = session,
          inputId = "merge_a",
          choices = reactive_datanames(),
          selected = restoreInput(session$ns("merge_a"), selected_a)
        )

        selected_b <- isolate(input$merge_b)
        if (identical(selected_b, "")) selected <- restoreInput(session$ns("merge_b"), NULL)
        teal.widgets::updateOptionalSelectInput(
          session = session,
          inputId = "merge_b",
          choices = reactive_datanames(),
          selected = restoreInput(session$ns("merge_b"), selected_b)
        )
      })

      merge_a <- reactive(input$merge_a)
      merge_b <- reactive(input$merge_b)

      reactive({
        logger::log_info("re-merging transform")
        new_data <- within(
          data(),
          ANL <- dplyr::left_join(merge_a, merge_b),
          merge_a = tryCatch(as.name(merge_a()), error = function(e) as.name("DatasetA")),
          merge_b = tryCatch(as.name(merge_b()), error = function(e) as.name("DatasetA"))
        )
        teal.data::datanames(new_data) <- c(teal.data::datanames(new_data), "ANL")

        logger::log_info("Number of rows in ADSL: {nrow(data()[['ADSL']])}")

        new_data
      })
    })
  }
)

transformer_head <- teal_data_module(
  label = "Keep first half of datasets",
  ui = function(id) actionButton(NS(id, "button"), "Click to keep first half"),
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      button_count <- reactive({
        tryCatch(req(input$button), error = function(e) 0)
      })
      data_new <- reactive({
        data_out <- data()
        if (button_count() %% 2 == 0) {
          updateActionButton(session, "button", label = "Click to keep first half")
          data_out # do nothing
        } else {
          updateActionButton(session, "button", label = "Click to restore full dataset")
          Reduce(
            function(x, name) within(x, dat_ref <- head(dat_ref, n = NROW(dat_ref) / 2), dat_ref = as.name(name)),
            init = data_out,
            teal.data::datanames(data_out)
          )
        }
      })
    })
  }
)

transformer_error <- teal_data_module(
  label = "Add (optional) error",
  ui = function(id) checkboxInput(NS(id, "add_error"), "Add error?", value = FALSE),
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      add_error_rv <- reactive(tryCatch(req(input$add_error), error = function(e) FALSE))

      reactive({
        if (add_error_rv()) {
          data() |> within(stop("Manual error here"))
        } else {
          data()
        }
      })
    })
  }
)

my_transformers <- list(transformer_dummy, transformer_merge_anl, transformer_error, transformer_head, transformer_error)

# Transformer modules definition ----------------------------------------------

data <- teal_data_module(
  ui = function(id) {
    ns <- NS(id)
    tagList(
      textInput(ns("username"), label = "Username"),
      passwordInput(ns("password"), label = "Password"),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$submit, {
        data <- teal_data() |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- teal.data::rADSL
              ADTTE <- teal.data::rADTTE
              iris <- iris

              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            username = input$username,
            password = input$password
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("ADSL", "ADTTE", "iris", "CO2")
        data
      })
    })
  }
)


data <- teal_data() |>
  within(
    {
      logger::log_trace("Loading data")
      ADSL <- teal.data::rADSL
      ADTTE <- teal.data::rADTTE
      iris <- iris

      CO2 <- CO2
      factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
      CO2[factors] <- lapply(CO2[factors], as.character)
    }
  )
join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
teal.data::datanames(data) <- c("ADSL", "ADTTE", "iris", "CO2")

init(
  data = data,
  modules = list(
    modules(
      label = "Group",
      example_module(
        label = "example teal module (datanames = subset and ANL)",
        transformers = my_transformers,
        datanames = c("ADSL", "ADTTE", "ANL")
      ),
      example_module(
        label = "example teal module (datanames = subset and ANL)",
        transformers = my_transformers,
        datanames = c("ADSL", "ADTTE", "ANL")
      )
    ),
    example_module(
      label = "example teal module (datanames = subset and ANL)",
      transformers = my_transformers,
      datanames = c("ADSL", "ADTTE", "ANL")
    ),
    example_module(
      label = "example teal module (datanames = all)",
      transformers = my_transformers
    )
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L))
  )
) |>
  shiny::runApp()
```

</details>

Reply:
This is for tests. Thanks to this we can test almost everything through `srv_teal` (which should be the only exported  module). See here for example https://github.com/insightsengineering/teal/blob/191d378aad298d17e7675e59c7e558cc6613cbef/tests/testthat/test-module_teal.R#L129

Reply:
Gotcha!